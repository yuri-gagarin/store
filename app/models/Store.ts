import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;