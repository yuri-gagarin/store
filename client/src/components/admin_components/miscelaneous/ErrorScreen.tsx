import React from 'react'
import { Button, Dimmer, Header, Icon, Segment } from 'semantic-ui-react'
import axios from "axios";

type RetryLastReqFunc = <T> () => T
interface Props {
  errorMessage?: string;
  lastRequest?: () => Promise<void>;
}
const ErrorScreen: React.FC<Props> = ({ errorMessage, lastRequest }): JSX.Element => {

  const handleRetry = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (lastRequest) {
      lastRequest();
    }
  };
  
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active>
          <Header as='h2' icon inverted>
            <Icon name='exclamation circle' />
              An Error Occured!
            <Header.Subheader>Dimmer sub-header</Header.Subheader>
            <Button 
              id="errorScreenRetryButton"
              inverted 
              color="green" 
              content="Retry" 
              onClick={handleRetry} 
            />
          </Header>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default ErrorScreen