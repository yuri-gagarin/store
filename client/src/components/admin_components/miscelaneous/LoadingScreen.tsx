import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText