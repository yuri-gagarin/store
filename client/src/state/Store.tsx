import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

