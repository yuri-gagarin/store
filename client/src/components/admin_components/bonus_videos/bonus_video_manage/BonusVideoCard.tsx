import React, { useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import EditControls from "./EditControls"
// css imports //
import "./css/bonusVideoCard.css";
// actions and state /
import { deleteBonusVideo } from "../actions/APIBonusVideoActions";
import { setCurrentBonusVideo, clearCurrentBonusVideo } from "../actions/UIBonusVideoActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps } from "react-router-dom";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props extends RouteComponentProps {
  bonusVideo: IBonusVideoData;
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const BonusVideoCard: React.FC<Props> = ({ bonusVideo, state, dispatch, history }): JSX.Element => {
  const [ editing, setEditing ] = useState<boolean>(false);
  const baseUrl = "/admin/home/my_bonus_videos/manage"
  const { _id, description, youTubeURL, vimeoURL, createdAt, editedAt } = bonusVideo;

  const handleBonusVideoOpen = (e: React.MouseEvent<HTMLButtonElement>): void => {
    history.push(baseUrl + "/view");
  }
  const handleBonusVideoEdit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (!editing) {
      setCurrentBonusVideo(_id, dispatch, state);
      history.push(baseUrl + "/edit");
      setEditing(true);
    } else {
      clearCurrentBonusVideo(dispatch);
      history.push(baseUrl);
      setEditing(false);
    }
  }
  const handleBonusVideoDelete = (e: React.MouseEvent<HTMLButtonElement>): void => {
    deleteBonusVideo( _id, dispatch, state)
      .then((success) => {
        
      })
  }

  return (
    <React.Fragment>
      <Grid.Row style={{ padding: "0.5em", marginTop: "1em" }}>
        <div className="bonusVideoManageDesc">
          <p><strong>Description:</strong> {description} </p>
          <p><strong>YouTube URL:</strong> {youTubeURL} </p>
          <p><strong>Vimeo URL:</strong> {vimeoURL} </p>
          <p><span>Uploaded At: </span>{ConvertDate.international(createdAt)}</p>
        </div> 
        <div className="bonusVideoManageCtrls">
          <Button 
            inverted
            color="green"
            className="bonusVideoCardBtn" 
            content="Open" 
            onClick={handleBonusVideoOpen} 
          />
          <Button 
            inverted
            color="orange"
            className="bonusVideoCardBtn" 
            content="Edit" 
            onClick={handleBonusVideoEdit}  
          />
          <Button 
            inverted
            color="red"
            className="bonusVideoCardBtn" 
            content="Delete" 
            onClick={handleBonusVideoDelete}
          />
        </div> 
      </Grid.Row>
      {  editing ?  <EditControls handleBonusVideoEdit={handleBonusVideoEdit}/>: null }
    </React.Fragment>
    
  );
};

export default withRouter(BonusVideoCard);