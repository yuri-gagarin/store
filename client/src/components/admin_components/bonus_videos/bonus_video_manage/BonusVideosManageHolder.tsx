import React, { useEffect, useContext  } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import BonusVideoFormHolder from "../forms/BonusVideosFormHolder";
import BonusVideoCard from "./BonusVideoCard";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
import ErrorScreen from "../../miscelaneous/ErrorScreen";
// actions and state //
import { getAllBonusVideos } from "../actions/APIBonusVideoActions";
import { Store } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";

interface Props extends RouteComponentProps {
};

const BonusVideosManageHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedBonusVideos , error } = state.bonusVideoState;
  // match route to conditionall render //
  const match = useRouteMatch("/admin/home/my_bonus_videos/manage");

  const handleBack = () => {
    history.goBack();
  };
  useEffect(() => {
    getAllBonusVideos(dispatch)
      .then((_) => {

      })
      .catch((_) => {
        // handle an error ? //
      })
  }, []); 
  useEffect(() => {
    console.log(loading)
  }, [ loading ])
  return (
    loading ? 
    <LoadingScreen />
    :
     ( error ? 
      <ErrorScreen lastRequest={ () => getAllBonusVideos(dispatch) } />
      :
      <Grid padded stackable columns={2}>
        <Route path={match?.url + "/edit"}> 
          <Grid.Row>
            <Grid.Column computer={12} tablet={6} mobile={16}>
              <h3>Editing Video: { state.bonusVideoState.currentBonusVideoData.description }</h3>
              <Button inverted color="green" content="Back" onClick={handleBack}></Button>
            </Grid.Column>
          </Grid.Row>
          <BonusVideoFormHolder />
        </Route>
        <Route exact path={match?.url}>
          <Grid.Row>
            <Grid.Column computer={12} tablet={8} mobile={16}>
            {
              loadedBonusVideos.map((bonusVideo) => {
                return (
                  <BonusVideoCard 
                    key={bonusVideo._id}
                    bonusVideo={bonusVideo}
                    state={state}
                    dispatch={dispatch}
                  />
                );
              })
            }
            </Grid.Column>
            <Grid.Column computer={4} tablet={8} mobile={16}>
              
            </Grid.Column>
          </Grid.Row>
        </Route>
      </Grid>
     )
  );
};

export default withRouter(BonusVideosManageHolder);