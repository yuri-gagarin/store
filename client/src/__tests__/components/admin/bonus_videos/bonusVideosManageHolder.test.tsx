import React from "react";
import { Grid } from "semantic-ui-react";
import moxios from "moxios";
// test dependencies
import { mount, ReactWrapper } from "enzyme";
import { act } from "react-dom/test-utils";
// routing //
import { MemoryRouter as Router } from "react-router-dom";
// components //
import BonusVideoManageHolder from "../../../../components/admin_components/bonus_videos/bonus_video_manage/BonusVideosManageHolder";
import BonusVideoCard from "../../../../components/admin_components/bonus_videos/bonus_video_manage/BonusVideoCard";
import ErrorScreen from "../../../../components/admin_components/miscelaneous/ErrorScreen";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
// helpers and state //
import { StateProvider } from "../../../../state/Store";
import BonusVideosManageHolder from "../../../../components/admin_components/bonus_videos/bonus_video_manage/BonusVideosManageHolder";

describe("BonusVideo Manage Holder Tests", () => {
  /*
  describe("Default Component state at first render", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;

    beforeAll( async () => {
      moxios.install();

      component = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_bonus_videos/manage"]}>
          <StateProvider>
            <BonusVideoManageHolder />
          </StateProvider>
        </Router>
      );

      await act( async () => {
        await moxios.stubRequest("/api/bonus_videos", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            bonusVideos: []
          }
        });
      });

    });
      afterAll(() => {
      moxios.uninstall();
    }); 
    
    it("Should correctly render", () => {
      const manageHolder = component.find(BonusVideosManageHolder);
      expect(manageHolder).toMatchSnapshot();
    });
    it("Should render a 'LoadingScreen' Component before an API call resolves", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1)
    });
  });
  
  // mock successful API call render tests //
  describe("State after a successful API call", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;
    const bonusVideos: IBonusVideoData[] = [];

    beforeAll( async () => {
      moxios.install();
      const vidData: IBonusVideoData = {
        _id: "1",
        youTubeURL: "url",
        vimeoURL: "url",
        description: "Something",
        createdAt: "1111"
      };
      bonusVideos.push(vidData);

      component = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_bonus_videos/manage"]}>
          <StateProvider>
            <BonusVideoManageHolder />
          </StateProvider>
        </Router>
      );
      
      await act( async () => {
        await moxios.stubRequest("/api/bonus_videos", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            bonusVideos: bonusVideos
          }
        });
      });
    });

    afterAll(() => {
      moxios.uninstall();
    }); 

    it("Should correctly render the initial Loading Screen", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not render the initial Loading Screen after the API call", async () => {
      component.update();
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should NOT render the ErrorScreen Component", () => {
      const errorScreenComponent = component.find(ErrorScreen);
      expect(errorScreenComponent.length).toEqual(0);
    });
    it("Should render the correct BonusVideoManageHolder Component", () => {
      const bonusVideoManageHolderComp = component.find(BonusVideoManageHolder).find(Grid);
      expect(bonusVideoManageHolderComp.length).toEqual(1);
      expect(bonusVideoManageHolderComp).toMatchSnapshot();
    });
    it("Should render correct number of BonusVideoCard components", () => {
      const bonusVideoCards = component.find(BonusVideoCard);
      expect(bonusVideoCards.length).toEqual(bonusVideos.length);
    });
  });
  
  // END mock successfull API call render tests //
  // TEST mock ERROR API call render tests //
  */
  describe("State after a Error in API call", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;
    const bonusVideos: IBonusVideoData[] = [];

    beforeAll(async () => {
      await act(async () => {
        moxios.install();
        component = await mount(
          <Router keyLength={0} initialEntries={["/admin/home/my_bonus_videos/manage"]}>
            <StateProvider>
              <BonusVideoManageHolder />
            </StateProvider>
          </Router>
        );
        moxios.stubRequest("/api/bonus_videos", {
          status: 500,
          response: {
            responseMsg: "Error here",
            error: new Error("API Call Error")
          }
        });
      });
      moxios.uninstall();
    });
    
    it("Should correctly render the initial Loading Screen", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not render the initial Loading Screen after an  API call", () => {
      component.update();
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should render the ErrorScreen Component", () => {
      const errorScreenComponent = component.find(ErrorScreen);
      expect(errorScreenComponent.length).toEqual(1);
    });
    it("Should NOT render the BonusVideoManageHolder Component", () => {
      const bonusVideoManageHolderComp = component.find(BonusVideoManageHolder).find(Grid);
      expect(bonusVideoManageHolderComp.length).toEqual(0);
    });
    it("Should NOT render ANY BonusVideoCard components", () => {
      const bonusVideoCards = component.find(BonusVideoCard);
      expect(bonusVideoCards.length).toEqual(0);
    });
    it("Should have a retry BonusVideo API call Button", () => {
      const retryButton = component.find("#errorScreenRetryButton");
      expect(retryButton.length).toEqual(2);
    });
    
    it("Should correctly re-dispatch the 'getBonusVideos' API request with the button click", async () => {
      const vidData: IBonusVideoData = {
        _id: "1",
        youTubeURL: "url",
        vimeoURL: "url",
        description: "Something",
        createdAt: "1111"
      };
      bonusVideos.push(vidData);
      await act( async () => {
        moxios.install();
        await moxios.stubRequest("/api/bonus_videos", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            bonusVideos: bonusVideos
          }
        });
        const retryButton = component.find("#errorScreenRetryButton");
        retryButton.at(0).simulate("click");  
      });
      // update component and assert correct rendering //
      component.update();
      const errorScreen = component.find(ErrorScreen);
      const bonusVideoManageHolderComp = component.find(BonusVideoManageHolder);
      expect(errorScreen.length).toEqual(0);
      expect(bonusVideoManageHolderComp.length).toEqual(1);
    });
    it("Should render correct number of 'BonusVideoCard' Components", () => {
      const bonusVideoCards = component.find(BonusVideoManageHolder).find(BonusVideoCard);
      expect(bonusVideoCards.length).toEqual(bonusVideos.length);
    });
    
  });
  
  // END mock successfull API call tests //
});