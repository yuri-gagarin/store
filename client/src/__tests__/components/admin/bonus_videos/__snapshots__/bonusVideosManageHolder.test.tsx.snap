// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusVideo Manage Holder Tests Default Component state at first render Should correctly render 1`] = `
<withRouter(BonusVideosManageHolder)>
  <BonusVideosManageHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_bonus_videos/manage",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_bonus_videos/manage",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_bonus_videos/manage",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <LoaderExampleText>
      <div
        id="generalLoadingScreen"
        style={
          Object {
            "height": "100vh",
          }
        }
      >
        <Segment
          style={
            Object {
              "height": "100vh",
            }
          }
        >
          <div
            className="ui segment"
            style={
              Object {
                "height": "100vh",
              }
            }
          >
            <Dimmer
              active={true}
              inverted={true}
            >
              <DimmerInner
                active={true}
                inverted={true}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": null,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": null,
                      }
                    }
                  >
                    <div
                      className="ui active transition visible inverted dimmer"
                      onClick={[Function]}
                    >
                      <div
                        className="content"
                      >
                        <Loader>
                          <div
                            className="ui text loader"
                          >
                            Please Wait
                          </div>
                        </Loader>
                      </div>
                    </div>
                  </RefFindNode>
                </Ref>
              </DimmerInner>
            </Dimmer>
          </div>
        </Segment>
      </div>
    </LoaderExampleText>
  </BonusVideosManageHolder>
</withRouter(BonusVideosManageHolder)>
`;

exports[`BonusVideo Manage Holder Tests State after a successful API call Should render the correct BonusVideoManageHolder Component 1`] = `
<withRouter(BonusVideosManageHolder)>
  <BonusVideosManageHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_bonus_videos/manage",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_bonus_videos/manage",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_bonus_videos/manage",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <Grid
      columns={2}
      padded={true}
      stackable={true}
    >
      <div
        className="ui stackable padded two column grid"
      >
        <Route
          path="/admin/home/my_bonus_videos/manage/edit"
        />
        <Route
          exact={true}
          path="/admin/home/my_bonus_videos/manage"
        >
          <GridRow>
            <div
              className="row"
            >
              <GridColumn
                computer={12}
                mobile={16}
                tablet={8}
              >
                <div
                  className="twelve wide computer sixteen wide mobile eight wide tablet column"
                >
                  <withRouter(BonusVideoCard)
                    bonusVideo={
                      Object {
                        "_id": "1",
                        "createdAt": "1111",
                        "description": "Something",
                        "vimeoURL": "url",
                        "youTubeURL": "url",
                      }
                    }
                    dispatch={[Function]}
                    key="1"
                    state={
                      Object {
                        "bonusVideoState": Object {
                          "bonusVideoFormOpen": false,
                          "currentBonusVideoData": Object {
                            "_id": "",
                            "createdAt": "",
                            "description": "",
                            "vimeoURL": "",
                            "youTubeURL": "",
                          },
                          "error": null,
                          "loadedBonusVideos": Array [
                            Object {
                              "_id": "1",
                              "createdAt": "1111",
                              "description": "Something",
                              "vimeoURL": "url",
                              "youTubeURL": "url",
                            },
                          ],
                          "loading": false,
                          "responseMsg": "All Ok",
                        },
                        "productState": Object {
                          "currentProductData": Object {
                            "_id": "",
                            "createdAt": "",
                            "description": "",
                            "details": "",
                            "images": Array [],
                            "name": "",
                            "price": "",
                          },
                          "error": null,
                          "loadedProducts": Array [],
                          "loading": false,
                          "productFormOpen": false,
                          "responseMsg": "",
                        },
                        "serviceState": Object {
                          "currentServiceData": Object {
                            "_id": "",
                            "createdAt": "",
                            "description": "",
                            "images": Array [],
                            "name": "",
                            "price": "",
                          },
                          "error": null,
                          "loadedServices": Array [],
                          "loading": false,
                          "responseMsg": "",
                          "serviceFormOpen": false,
                        },
                        "storeItemState": Object {
                          "currentStoreItemData": Object {
                            "_id": "",
                            "categories": Array [],
                            "createdAt": "",
                            "description": "",
                            "details": "",
                            "images": Array [],
                            "name": "",
                            "price": "",
                            "storeId": "",
                            "storeName": "",
                          },
                          "error": null,
                          "loadedStoreItems": Array [],
                          "loading": false,
                          "numberOfItems": 0,
                          "responseMsg": "",
                          "storeItemFormOpen": false,
                        },
                        "storeState": Object {
                          "currentStoreData": Object {
                            "_id": "",
                            "createdAt": "",
                            "description": "",
                            "images": Array [],
                            "title": "",
                          },
                          "error": null,
                          "loadedStores": Array [],
                          "loading": false,
                          "responseMsg": "",
                          "storeFormOpen": false,
                        },
                      }
                    }
                  >
                    <BonusVideoCard
                      bonusVideo={
                        Object {
                          "_id": "1",
                          "createdAt": "1111",
                          "description": "Something",
                          "vimeoURL": "url",
                          "youTubeURL": "url",
                        }
                      }
                      dispatch={[Function]}
                      history={
                        Object {
                          "action": "POP",
                          "block": [Function],
                          "canGo": [Function],
                          "createHref": [Function],
                          "entries": Array [
                            Object {
                              "hash": "",
                              "pathname": "/admin/home/my_bonus_videos/manage",
                              "search": "",
                              "state": undefined,
                            },
                          ],
                          "go": [Function],
                          "goBack": [Function],
                          "goForward": [Function],
                          "index": 0,
                          "length": 1,
                          "listen": [Function],
                          "location": Object {
                            "hash": "",
                            "pathname": "/admin/home/my_bonus_videos/manage",
                            "search": "",
                            "state": undefined,
                          },
                          "push": [Function],
                          "replace": [Function],
                        }
                      }
                      location={
                        Object {
                          "hash": "",
                          "pathname": "/admin/home/my_bonus_videos/manage",
                          "search": "",
                          "state": undefined,
                        }
                      }
                      match={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/admin/home/my_bonus_videos/manage",
                          "url": "/admin/home/my_bonus_videos/manage",
                        }
                      }
                      state={
                        Object {
                          "bonusVideoState": Object {
                            "bonusVideoFormOpen": false,
                            "currentBonusVideoData": Object {
                              "_id": "",
                              "createdAt": "",
                              "description": "",
                              "vimeoURL": "",
                              "youTubeURL": "",
                            },
                            "error": null,
                            "loadedBonusVideos": Array [
                              Object {
                                "_id": "1",
                                "createdAt": "1111",
                                "description": "Something",
                                "vimeoURL": "url",
                                "youTubeURL": "url",
                              },
                            ],
                            "loading": false,
                            "responseMsg": "All Ok",
                          },
                          "productState": Object {
                            "currentProductData": Object {
                              "_id": "",
                              "createdAt": "",
                              "description": "",
                              "details": "",
                              "images": Array [],
                              "name": "",
                              "price": "",
                            },
                            "error": null,
                            "loadedProducts": Array [],
                            "loading": false,
                            "productFormOpen": false,
                            "responseMsg": "",
                          },
                          "serviceState": Object {
                            "currentServiceData": Object {
                              "_id": "",
                              "createdAt": "",
                              "description": "",
                              "images": Array [],
                              "name": "",
                              "price": "",
                            },
                            "error": null,
                            "loadedServices": Array [],
                            "loading": false,
                            "responseMsg": "",
                            "serviceFormOpen": false,
                          },
                          "storeItemState": Object {
                            "currentStoreItemData": Object {
                              "_id": "",
                              "categories": Array [],
                              "createdAt": "",
                              "description": "",
                              "details": "",
                              "images": Array [],
                              "name": "",
                              "price": "",
                              "storeId": "",
                              "storeName": "",
                            },
                            "error": null,
                            "loadedStoreItems": Array [],
                            "loading": false,
                            "numberOfItems": 0,
                            "responseMsg": "",
                            "storeItemFormOpen": false,
                          },
                          "storeState": Object {
                            "currentStoreData": Object {
                              "_id": "",
                              "createdAt": "",
                              "description": "",
                              "images": Array [],
                              "title": "",
                            },
                            "error": null,
                            "loadedStores": Array [],
                            "loading": false,
                            "responseMsg": "",
                            "storeFormOpen": false,
                          },
                        }
                      }
                    >
                      <GridRow
                        style={
                          Object {
                            "marginTop": "1em",
                            "padding": "0.5em",
                          }
                        }
                      >
                        <div
                          className="row"
                          style={
                            Object {
                              "marginTop": "1em",
                              "padding": "0.5em",
                            }
                          }
                        >
                          <div
                            className="bonusVideoManageDesc"
                          >
                            <p>
                              <strong>
                                Description:
                              </strong>
                               
                              Something
                               
                            </p>
                            <p>
                              <strong>
                                YouTube URL:
                              </strong>
                               
                              url
                               
                            </p>
                            <p>
                              <strong>
                                Vimeo URL:
                              </strong>
                               
                              url
                               
                            </p>
                            <p>
                              <span>
                                Uploaded At: 
                              </span>
                              1111-Jan-1
                            </p>
                          </div>
                          <div
                            className="bonusVideoManageCtrls"
                          >
                            <Button
                              as="button"
                              className="bonusVideoCardBtn"
                              color="green"
                              content="Open"
                              inverted={true}
                              onClick={[Function]}
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui green inverted button bonusVideoCardBtn"
                                    >
                                      Open
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui green inverted button bonusVideoCardBtn"
                                      >
                                        Open
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui green inverted button bonusVideoCardBtn"
                                    onClick={[Function]}
                                  >
                                    Open
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                            <Button
                              as="button"
                              className="bonusVideoCardBtn"
                              color="orange"
                              content="Edit"
                              inverted={true}
                              onClick={[Function]}
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui orange inverted button bonusVideoCardBtn"
                                    >
                                      Edit
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui orange inverted button bonusVideoCardBtn"
                                      >
                                        Edit
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui orange inverted button bonusVideoCardBtn"
                                    onClick={[Function]}
                                  >
                                    Edit
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                            <Button
                              as="button"
                              className="bonusVideoCardBtn"
                              color="red"
                              content="Delete"
                              inverted={true}
                              onClick={[Function]}
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui red inverted button bonusVideoCardBtn"
                                    >
                                      Delete
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui red inverted button bonusVideoCardBtn"
                                      >
                                        Delete
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui red inverted button bonusVideoCardBtn"
                                    onClick={[Function]}
                                  >
                                    Delete
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                          </div>
                        </div>
                      </GridRow>
                    </BonusVideoCard>
                  </withRouter(BonusVideoCard)>
                </div>
              </GridColumn>
              <GridColumn
                computer={4}
                mobile={16}
                tablet={8}
              >
                <div
                  className="four wide computer sixteen wide mobile eight wide tablet column"
                />
              </GridColumn>
            </div>
          </GridRow>
        </Route>
      </div>
    </Grid>
  </BonusVideosManageHolder>
</withRouter(BonusVideosManageHolder)>
`;
